app/models/prolong_token_wizard.rb

```diff
@@ -0,0 +1,61 @@
+class ProlongTokenWizard < ApplicationRecord
+  belongs_to :token
+
+  enum status: {
+    owner: 0,
+    project_purpose: 10,
+    contacts: 20,
+    finished: 100,
+    requires_update: 110,
+    updates_requested: 120,
+    updates_refused: 130,
+    prolonged: 200
+  }
+
+  validates :owner, presence: true, if: -> { after?(:owner) }
+  validates :project_purpose, inclusion: { in: [true, false] }, if: -> { after?(:project_purpose) }
+  validates :contact_technique, inclusion: { in: [true, false] }, if: -> { after?(:contacts) }
+  validates :contact_metier, inclusion: { in: [true, false] }, if: -> { after?(:contacts) }
+
+  scope :not_finished, -> { where('status IS NULL OR status NOT IN (?)', [ProlongTokenWizard.statuses[:updates_refused], ProlongTokenWizard.statuses[:prolonged]]) }
+
+  def finish!
+    return if after?(:finished)
+
+    if should_prolong_token?
+      prolong!
+    else
+      requires_update!
+    end
+  end
+
+  def new?
+    status.nil?
+  end
+
+  def renew?
+    after?(:updates_refused)
+  end
+
+  def after?(kind)
+    status.present? && ProlongTokenWizard.statuses[kind] <= ProlongTokenWizard.statuses[status]
+  end
+
+  def should_prolong_token?
+    owner != 'not_in_charge' &&
+      project_purpose &&
```

Comment:
```suggestion
project_purpose.present? &&
```

Plus explicite imo

Possiblement faire pareil avec les 2 lignes suivantes (moins important tho)

Reply:
Y avait un `keep_` avant mais on m'a demandé de retirer car inutile.

Donc j'suis ok avec toi, mais j'ai pas vraiment d'idée concrète là dessus :D

Reply:
Maybe `same_` ou `persisted_` ?

Reply:
ca revient au `keep_` qu'il y avait avant qui était trop verbeux.

Reply:
meme le `keep_` je trouve ça mieux que rien mais osef si @skelz0r a demandé de retirer je vais pas râler :+1: